src/StreamJsonRpc/JsonRpc.cs

```diff
@@ -1333,6 +1343,15 @@ private void OnJsonRpcDisconnected(JsonRpcDisconnectedEventArgs eventArgs)
 
         private async Task ReadAndHandleRequestsAsync()
         {
+            lock (this.syncObject)
```

Comment:
so, this relies on the behavior of Task.Run that it will run func<Task> in another thread? since lock statement allows re-entrance on the same thread. 

if that is guaranteed, seems fine. otherwise, it would still have a chance to have the issue.

Reply:
Yes, `Task.Run`'s documented behavior is to schedule work on `TaskScheduler.Default` (the thread pool). It never inlines the delegate passed to it.